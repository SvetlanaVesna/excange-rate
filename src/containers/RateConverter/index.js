import React from 'react'
import { connect } from 'react-redux'

import {
	compose,
	withStateHandlers,
	withHandlers,
	lifecycle,
	withProps,
} from 'recompose'

import * as select from '../../reducers/selectors'

import {
	ratePollStartAction,
	ratePollStopAction,
	selectWallet,
	exchange,
} from '../../actions'

import { formatter, checkInputValue } from '../../utils'

import RateConvertComponent from './component'

const RateConvertContainer = compose(
	connect(
		state => ({
			sourceCurrency: select.getSelectedSourceCurrency(state),
			targetCurrency: select.getSelectedTargetCurrency(state),
			convertIndex: select.getCurrenciesRelation(state),
			isFetching: select.isRatesFetching(state),
			userWallets: select.getAllUserWallets(state),
			selectedWallet: select.getSourceUserWallet(state),
			targetUserWallet: select.getTargetUserWallet(state),
		}),
		dispatch => ({ dispatch }),
	),
	withStateHandlers(
		() => ({
			valueToConvert: '',
			error: false,
		}),
		{
			setStateValue: () => val => val,
		},
	),
	withProps(({ valueToConvert, convertIndex, targetCurrency }) => ({
		resultValue: formatter(targetCurrency, valueToConvert * convertIndex),
	})),
	withHandlers({
		setValueToConvert: ({
			setStateValue,
			selectedWallet: { content },
		}) => val => setStateValue(checkInputValue(content, val)),

		ratePollStopAction: ({ dispatch }) => () => dispatch(ratePollStopAction()),

		selectWallet: ({
			dispatch,
			userWallets,
			valueToConvert,
			setStateValue,
		}) => (direction, wallet) => {
			dispatch(selectWallet(direction, wallet))
			const hasError = checkInputValue(
				userWallets[wallet].content,
				valueToConvert,
			)
			setStateValue(hasError)
		},

		exchange: ({
			dispatch,
			valueToConvert,
			resultValue,
			selectedWallet: { content },
			setStateValue,
		}) => () => {
			const hasError = checkInputValue(content, valueToConvert)
			if (!hasError.error) dispatch(exchange(valueToConvert, resultValue))
			else setStateValue(hasError)
		},
	}),
	lifecycle({
		componentDidMount() {
			const { dispatch } = this.props
			dispatch(ratePollStartAction())
		},
	}),
)(RateConvertComponent)

const RateConvert = () => <RateConvertContainer />

export default RateConvert
