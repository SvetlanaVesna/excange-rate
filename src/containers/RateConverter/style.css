.wrapper, .footer {
    display: flex;
    flex-flow: row wrap;
    font-weight: bold;
    text-align: center;
}

.wrapper > *, .footer > * {
    padding: 10px;
    flex: 1 100%;
}

.header {
    background: deepskyblue;
}

.footer {
    margin-top: 10px;
    background: transparent;
    -webkit-box-shadow: 10px 10px 20px -4px rgba(0,0,0,0.29);
    -moz-box-shadow: 10px 10px 20px -4px rgba(0,0,0,0.29);
    box-shadow: 10px 10px 20px -4px rgba(0,0,0,0.29);
}

.aside-1 {
    background: transparent;
    -webkit-box-shadow: 10px 10px 20px -4px rgba(0,0,0,0.29);
    -moz-box-shadow: 10px 10px 20px -4px rgba(0,0,0,0.29);
    box-shadow: 10px 10px 20px -4px rgba(0,0,0,0.29);
}
.main {
    text-align: center;
    font-size: large;
}

.aside-2 {
    background: transparent;
    -webkit-box-shadow: 10px 10px 20px -4px rgba(0,0,0,0.29);
    -moz-box-shadow: 10px 10px 20px -4px rgba(0,0,0,0.29);
    box-shadow: 10px 10px 20px -4px rgba(0,0,0,0.29);
}

@media all and (min-width: 600px) {
    .aside { flex: 1 auto; }
}

@media all and (min-width: 800px) {
    .main    { flex: 3 0px; }
    .aside-1 { order: 1; }
    .main    { order: 2; }
    .aside-2 { order: 3; }
    .footer  { order: 4; }
}

body {
    padding: 2em;
}
.convert_value_output{
    display: inline-block;
    -webkit-box-sizing: content-box;
    -moz-box-sizing: content-box;
    box-sizing: content-box;
    padding: 10px 20px;
    border: 1px solid #b7b7b7;
    -webkit-border-radius: 3px;
    border-radius: 3px;
    font: normal 16px/normal "Times New Roman", Times, serif;
    color: black;
    -o-text-overflow: clip;
    text-overflow: clip;
    background: rgba(252,252,252,1);
    -webkit-box-shadow: 2px 2px 2px 0 rgba(0,0,0,0.2) inset;
    box-shadow: 2px 2px 2px 0 rgba(0,0,0,0.2) inset;
    text-shadow: 1px 1px 0 rgba(255,255,255,0.66) ;
    -webkit-transition: all 200ms cubic-bezier(0.42, 0, 0.58, 1);
    -moz-transition: all 200ms cubic-bezier(0.42, 0, 0.58, 1);
    -o-transition: all 200ms cubic-bezier(0.42, 0, 0.58, 1);
    transition: all 200ms cubic-bezier(0.42, 0, 0.58, 1);
}
.convert_value_input{
    display: inline-block;
    -webkit-box-sizing: content-box;
    -moz-box-sizing: content-box;
    box-sizing: content-box;
    padding: 10px 20px;
    border: 1px solid #b7b7b7;
    -webkit-border-radius: 3px;
    border-radius: 3px;
    font: normal 16px/normal "Times New Roman", Times, serif;
    color: rgba(0,142,198,1);
    -o-text-overflow: clip;
    text-overflow: clip;
    background: rgba(252,252,252,1);
    -webkit-box-shadow: 2px 2px 2px 0 rgba(0,0,0,0.2) inset;
    box-shadow: 2px 2px 2px 0 rgba(0,0,0,0.2) inset;
    text-shadow: 1px 1px 0 rgba(255,255,255,0.66) ;
    -webkit-transition: all 200ms cubic-bezier(0.42, 0, 0.58, 1);
    -moz-transition: all 200ms cubic-bezier(0.42, 0, 0.58, 1);
    -o-transition: all 200ms cubic-bezier(0.42, 0, 0.58, 1);
    transition: all 200ms cubic-bezier(0.42, 0, 0.58, 1);
}
.input_error_label{
    color: red;
}
.input_error{
    border-color: red;
}
.rate_select {
    padding:10px 20px;
    margin: 0;
    -webkit-border-radius:4px;
    -moz-border-radius:4px;
    border-radius:4px;
    -webkit-box-shadow: 0 3px 0 #ccc, 0 -1px #fff inset;
    -moz-box-shadow: 0 3px 0 #ccc, 0 -1px #fff inset;
    box-shadow: 0 3px 0 #ccc, 0 -1px #fff inset;
    background: #f8f8f8;
    color:#888;
    border:none;
    outline:none;
    display: inline-block;
    -webkit-appearance:none;
    -moz-appearance:none;
    cursor:pointer;
}

label {position:relative}
label:after {
    content:'<>';
    font:normal 16px/normal "Times New Roman", Times, serif;
    color:#aaa;
    -webkit-transform:rotate(90deg);
    -moz-transform:rotate(90deg);
    -ms-transform:rotate(90deg);
    transform:rotate(90deg);
    right:2px; top:2px;
    padding:0 0 2px;
    border-bottom:1px solid #ddd;
    position:absolute;
    pointer-events:none;
}
label:before {
    content:'';
    right:0px; top:0px;
    width:20px; height:20px;
    background:#f8f8f8;
    position:absolute;
    pointer-events:none;
    display:block;
}