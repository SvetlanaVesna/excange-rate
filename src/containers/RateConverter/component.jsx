import React from 'react'

import Footer from './parts/footer'
import Header from './parts/header'

import WalletSelectionComponent from './parts/walletSelection'
import CurrencyInputComponent from './parts/currencyInput'
import TargetWalletComponent from './parts/targetWallet'

import './style.css'

const RateConverter = ({
	isFetching,

	sourceCurrency,
	targetCurrency,

	valueToConvert,
	setValueToConvert,

	resultValue,
	convertIndex,
	error,

	userWallets,
	selectedWallet,
	targetUserWallet,

	selectWallet,
	exchange,

	ratePollStopAction,
}) => (
	<div className="wrapper">
		<Header ratePollStopAction={ratePollStopAction} />
		<WalletSelectionComponent
			selectWallet={selectWallet}
			selectedWallet={selectedWallet}
			userWallets={userWallets}
		/>
		{isFetching && <h1>LOADING...</h1>}
		<CurrencyInputComponent
			error={error}
			setValueToConvert={setValueToConvert}
			valueToConvert={valueToConvert}
			sourceCurrency={sourceCurrency}
		/>
		{!isFetching && (
			<TargetWalletComponent
				resultValue={resultValue}
				selectWallet={selectWallet}
				targetCurrency={targetCurrency}
				userWallets={userWallets}
				exchange={exchange}
				error={error}
			/>
		)}
		{!isFetching && (
			<Footer
				selectedWallet={selectedWallet}
				sourceCurrency={sourceCurrency}
				targetUserWallet={targetUserWallet}
				targetCurrency={targetCurrency}
			/>
		)}
		<div>
			<div>
				1 {sourceCurrency} = {convertIndex} {targetCurrency}
			</div>
			<div>
				1 {targetCurrency} = {1 / convertIndex} {sourceCurrency}
			</div>
		</div>
	</div>
)

export default RateConverter
