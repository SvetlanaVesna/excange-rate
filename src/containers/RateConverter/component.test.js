import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import RateConverter from './index'
Enzyme.configure({ adapter: new Adapter() })

describe('<RateConverter />', () => {
	it('snapshot test for RateConverter component', () => {
		const eurRate = 0.879981
		const gbpRate = 0.782398
		const resultIndex = eurRate / gbpRate
		const inputValue = 12
		const resultValue = inputValue * resultIndex

		const props = {
			availableRates: ['GBP', 'EUR', 'USD'],
			sourceCurrency: 'GBP',
			targetCurrency: 'EUR',
			valueToConvert: 12,
			setValueToConvert: jest.fn(),
			selectCurrency: jest.fn(),
			resultValue: resultValue,
			ratePollStopAction: jest.fn(),
			convertIndex: resultIndex,
			error: false,
			isFetching: false,

			userWallets: {
				USD: {
					content: 58,
					currency: 'USD',
				},
				EUR: {
					content: 0,
					currency: 'EUR',
				},
				GBP: {
					content: 2,
					currency: 'GBP',
				},
			},
			selectedWallet: { currency: 'GBP', content: 2 },
			targetUserWallet: { currency: 'USD', content: 58 },

			selectWallet: jest.fn(),
			exchange: jest.fn(),
		}

		const component = shallow(<RateConverter {...props} />)

		expect(component).toMatchSnapshot()
	})
})
