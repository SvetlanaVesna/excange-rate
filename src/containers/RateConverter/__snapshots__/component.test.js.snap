// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RateConverter /> snapshot test for RateConverter component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RateConvert
    availableRates={
      Array [
        "GBP",
        "EUR",
        "USD",
      ]
    }
    convertIndex={1.1247229670832493}
    error={false}
    exchange={[MockFunction]}
    isFetching={false}
    ratePollStopAction={[MockFunction]}
    resultValue={13.496675604998991}
    selectCurrency={[MockFunction]}
    selectWallet={[MockFunction]}
    selectedWallet={
      Object {
        "content": 2,
        "currency": "GBP",
      }
    }
    setValueToConvert={[MockFunction]}
    sourceCurrency="GBP"
    targetCurrency="EUR"
    targetUserWallet={
      Object {
        "content": 58,
        "currency": "USD",
      }
    }
    userWallets={
      Object {
        "EUR": Object {
          "content": 0,
          "currency": "EUR",
        },
        "GBP": Object {
          "content": 2,
          "currency": "GBP",
        },
        "USD": Object {
          "content": 58,
          "currency": "USD",
        },
      }
    }
    valueToConvert={12}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
